controller/main.go

```diff
@@ -41,22 +44,27 @@ type service interface {
 }
 
 type controller struct {
-	client service
-	pools  *config.Pools
-	ips    *allocator.Allocator
+	client                  service
+	pools                   *config.Pools
+	ips                     *allocator.Allocator
+	generatePoolStatusEvent func(string)
```

Comment:
All the opposite, the logic is already very simple. It only compares the allocated pools before and after calling convergenceBalancer (if in line 100) and it only does it when there was a change in the status of the service. Note that none of the existing functions are modified.

On the other hand,  the allocator lacks information about when a change is really a change. Note that the Assign func of the allocator is also used to verify the LB IPs of services. So, if we simply generate the status event from the assign and unassign funcs (similar to the Prometheus metrics), excessive events will be generated. For instance, if a service with LB IPs requests a different pool, the current logic first calls assign (which always calls unassign) to check the existing IPs. Then, the pool is cleared out because the pool requested is different and finally the new pool is assigned. So, there will be 4 events for this case. Even a service with LB IPs without changes will generate 2 unneeded events. 

I believe this is simple enough and uses the existing DeepEqual checks to generate events only when needed. It follows the changes on the status of services.

Reply:
changed the name as suggested

Reply:
I don't agree with this. Nothing complex is being done here. And of course, we need to have some knowledge of the implementation, starting with the fact that we are creating a custom event. I understand your point about readability and maintainability but it doesn't apply here. This logic is not invasive and it is very simple.

Reply:
You don't need to know it's a custom event, as you have a "poolChanged" call back, which the caller ignores the semantics.
Here we are leveraging a side effect (the pools associated to the service changed) to assume that the pool changed, which is not straightforward.

Reply:
that means that the service was allocated (in the allocator). It is normal make that assumption. This follows the same logic of was/isServiceAllocated

Reply:
we send a reprocess all event when a pool is associated with a service. We are doing exactly the same assumption

Reply:
I agree with @fedepaol here, if it's possible to not make any assumptions based on the (current) behavior between the different components but rather only at the point that actually controls the allocations it's preferable over saving a few cycles (assuming we're not talking about something drastic). I feel that it'd be easier to understand/debug and we won't have to think about it when adding new code (that might be related to pools).

Reply:
I don't agree. Like I said, this is very simple and relies on a core principle which is finding a pool for a service. This is the main component alloc structure during the assign. Besides, I'm not introducing new assumption. We are already assuming that a reprocessallevent must be triggered when a pool is allocated. The assumption of a pool change when this is allocated/deallocated to a service is the key principle of the Balancer. It will be very rare to change this logic one day.

Reply:
This is very simple, but it can be simpler. When we have a bug or something to change, I don't want to come here and have to think about prevpool / newpool and think why we are calling the callback and why something it's not working.

And, as I wrote above, simpler is, we touch the pool, we call the callback. This doesn't require any thinking at all.
If the fact that assign / unassign would call the callback too many times is a concern, we should optimize those (specifically, assign because that's where the unnecessary updates are).

Reply:
Besides the issue with the amount of calls to assign/unassign, another problem I see with the allocator is that it doesn't always reflect the actual state of the allocation based on the status of the service. In fact, I believe the correct place to generate a pool status event is after updating the status of the service. Only then, we can assure that something has changed. For instance, if a svc with valid IPs requests a different pool, this new pool will be assigned even if the call to update the service status fail. The previous allocation is not restored. This leads again to excessive calls to assign/unassign (new pool to old pool to new pool again) in the next iteration. And until the service status is not updated, nothing has really changed. If I'm not mistaken, speakers check the LB ips to make decision.

So, having the logic one lever higher not only improves efficiency by lowering the amount of events sent to the status controller but also reduces the contention to the channel that protects the listener. Additionally, moving the logic after updating the service status will truly reflect the state of the allocation seeing from what the services have. And it is not difficult to understand, it's very simple. A comment would be enough.

I also think that there is a bug when the status call fails for a service with valid IPs requesting an invalid one. The allocation will be deleted and in the next iteration, the wasAllocated var will be false. The reprocessall event won't be generated. Something to address. Perhaps restoring the previous allocation.

Anyway, I see this discussion is not going anywhere. I totally understand if we decide to put this on hold and hand it over to a person who understand/agrees more the performance/correctness/simplicity tradeoffs and code style use in this project.

Reply:
Leaving the unnecessary events aside (which can be addressed in the allocator), the fact that the speaker is already relying on a side effect doesn't necessarily mean that relying on side effects is the right thing to do.

As you said we have different opinions on what the right thing to do is. I'll ask @oribon 's feedback again, maybe I am over simplifying.

Reply:
I will also reply to the last considerations and leave space to @oribon 

> Besides the issue with the amount of calls to assign/unassign, another problem I see with the allocator is that it doesn't always reflect the actual state of the allocation based on the status of the service. In fact, I believe the correct place to generate a pool status event is after updating the status of the service. Only then, we can assure that something has changed. For instance, if a svc with valid IPs requests a different pool, this new pool will be assigned even if the call to update the service status fail. The previous allocation is not restored. This leads again to excessive calls to assign/unassign (new pool to old pool to new pool again) in the next iteration. And until the service status is not updated, nothing has really changed. If I'm not mistaken, speakers check the LB ips to make decision.
> 

What we care about is exactly the content of the pool counters, not if the ips are assigned to a service or not. If we took an ip away from a pool, but we fail to assign it to a service, it still make sense to count that IP as used. The status of the available IPs must be reflect what IPs are available. And the best place to understand when those counters change is, where those counters change.


> So, having the logic one lever higher not only improves efficiency by lowering the amount of events sent to the status controller but also reduces the contention to the channel that protects the listener. Additionally, moving the logic after updating the service status will truly reflect the state of the allocation seeing from what the services have. And it is not difficult to understand, it's very simple. A comment would be enough.
> 

I already mentioned that I prefer the version I am advocating despite the a higher number of generated events (which again, can be addressed in the allocator to some extent) and a penalty in the performances. The other side of the channel is in a select that writes to a worker queue, so it won't be an issue. And to add more, multiple events written in the worker queue are squashed.
So, leaving performances aside, I already wrote above that the pool status must reflect the content of the counters.

> I also think that there is a bug when the status call fails for a service with valid IPs requesting an invalid one. The allocation will be deleted and in the next iteration, the wasAllocated var will be false. The reprocessall event won't be generated. Something to address. Perhaps restoring the previous allocation.

And this is another example on why we shouldn't rely on side effects but be straightforward, when possible.

> 
> Anyway, I see this discussion is not going anywhere. I totally understand if we decide to put this on hold and hand it over to a person who understand/agrees more the performance/correctness/simplicity tradeoffs and code style use in this project.

Reply:
I don't understand why you say this is a side effect. The allocator is simply an internal database that is relevant only to the controller (wiped out if the pod restarts). The information that really matters and used by other components is the status of services. If the update fails, the info from the allocator should be reverted to avoid the mentioned bug and have consistency between these 2 parts.

Anyway, the above can be fixed but the question about which approach to take remains. Don't take me wrong, I'm not saying that what you're proposing is not valid. It's not efficient and in my opinion, not 100% correct either, but it will work and if there was already an effort towards that direction, then it's fine. Here I believe that what I'm proposing has many advantages. It's efficient, simple, easy to understand and follows the info from the status of services (not yet, will need to move the logic after the update). It won't be consistent with the way the Prometheus metrics are taken though.

Of course, I don't want to impose anything. You are the maintainer and the final decision is yours. I'm here only to help. If you believe this is definitely a no go, then I'm happy to close this PR.

Reply:
We have clearly different views on what the source of truth should be. The status we want to represent is how many ips are available and how many are not. The allocator is the internal rappresentation of the pools, and as such, we should use it as the master (and we do, we change the counters when we have a new pool, and in this PR you fetch the numbers from the allocator and not by counting the ips assigned, service by service).

I might of course be wrong, but I am not convinced of the alternative approach. I need to do what I think it's best for the project in order for whoever contributes to have less friction in understanding or changing things.

And I think the most obvious way to do this is to intercept when those numbers change in the allocator. This has precedence over the performance considerations, that can (and should) be mitigated even with the version I am advocating for.

Reply:
Understood. I'm closing this pr now to give preference to your approach. Someone (or even myself) could use it as reference in the future.

Thanks for your comments.